lib/Dialect/Sim/Transforms/LowerDPIFunc.cpp

```diff
@@ -0,0 +1,176 @@
+//    %v = func.call @foo (%a, %0)
+//    func.return %v
+// }
+// hw.module @mod(..) {
+//    %result = sim.dpi.call @foo_wrapper(%a) clock %clock
+// }
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Sim/SimOps.h"
+#include "circt/Dialect/Sim/SimPasses.h"
+#include "circt/Support/Namespace.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "sim-lower-dpi-func"
+
+namespace circt {
+namespace sim {
+#define GEN_PASS_DEF_LOWERDPIFUNC
+#include "circt/Dialect/Sim/SimPasses.h.inc"
+} // namespace sim
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct LoweringState {
+  DenseMap<StringAttr, func::FuncOp> dpiFuncDeclMapping;
+  circt::Namespace nameSpace;
+};
+struct LowerDPIFuncPass : public sim::impl::LowerDPIFuncBase<LowerDPIFuncPass> {
+
+  LogicalResult lowerDPI();
+  LogicalResult lowerDPIFuncOp(sim::DPIFuncOp simFunc,
+                               LoweringState &loweringState,
+                               SymbolTable &symbolTable);
+  void runOnOperation() override;
+};
+
+} // namespace
+
+LogicalResult LowerDPIFuncPass::lowerDPIFuncOp(sim::DPIFuncOp simFunc,
+                                               LoweringState &loweringState,
+                                               SymbolTable &symbolTable) {
+  ImplicitLocOpBuilder builder(simFunc.getLoc(), simFunc);
+  auto moduleType = simFunc.getModuleType();
+
+  llvm::SmallVector<Type> dpiFunctionArgumentTypes;
+  for (auto arg : moduleType.getPorts()) {
+    // TODO: Support a non-integer type.
+    if (!arg.type.isInteger())
+      return simFunc->emitError()
+             << "non-integer type argument is unsupported now";
+
+    if (arg.dir == hw::ModulePort::Input)
+      dpiFunctionArgumentTypes.push_back(arg.type);
+    else
+      // Output must be passed by a reference.
+      dpiFunctionArgumentTypes.push_back(
+          LLVM::LLVMPointerType::get(arg.type.getContext()));
+  }
+
+  auto funcType = builder.getFunctionType(dpiFunctionArgumentTypes, {});
+  func::FuncOp func;
+
+  // Look up func.func by verilog name since the function name is equal to the
+  // symbol name in MLIR
+  if (auto verilogName = simFunc.getVerilogName()) {
+    func = symbolTable.lookup<func::FuncOp>(*verilogName);
+    // TODO: Check if function type matches.
+  }
+
+  // If a referred function is not in the same module, create an external
+  // function declaration.
+  if (!func) {
+    func = builder.create<func::FuncOp>(simFunc.getVerilogName()
+                                            ? *simFunc.getVerilogName()
```

Comment:
Nit: these two cases do not seem to be covered with tests

Reply:
Added in lower-dpi.mlir, thanks!